text/0014-monitoring.md

```diff
@@ -0,0 +1,66 @@
+# Collect and Expose Metrics
+
+## Summary
+
+[summary]: #summary
+
+The goal is to collect and expose structured metrics about the quality and status of the SLAM algorithm and the server that runs it.
+
+## Motivation
+
+[motivation]: #motivation
+
+### Dashboard for humans
+
+Human developers debug SLAM performance, and human operators need to monitor key metrics for the robot.
+Both can profit greatly from a nice dashboard that shows key cartographer metrics, such as a histogram of constraint quality.
+
+Today, developers often find themselves searching in log output, and the only way to monitor Cartographer is to run rviz at runtime.
+We need to provide a more structured way that also works for multiple instances or longer time scales.
+
+### Reference signal for robot control
+
+Some manufacturers would like to have an observable for SLAM performance so that they can make downstream decisions due to bad SLAM.
+Collected metrics could provide these reference signals with little intrusion in Cartographer code.
+
+## Approach
+
+[approach]: #approach
+
+The approach below is illustrated by a [prototype](https://github.com/googlecartographer/cartographer/pull/876).
+
+### Cartographer facade
+
+Inside the cartographer library, we establish a system-agnostic facade for collecting metrics.
+The facade is compatible with all important monitoring systems (Prometheus, Stackdriver, OpenCensus), but doesn't introduce build dependencies into the core cartographer lib.
+The facade lives in the cartographer::metrics namespace and includes these core classes:
+
+*   `Counter`: A single, always-increasing count of events, e.g. "number of loop iterations".
+*   `Gauge`: A value that changes both up and down, e.g. "memory used" or "open requests".
+*   `Histogram`: A thread-safe histogram with predefined buckets and constant memory usage.
+    It's different from `cartographer::common::Histogram`, which is thread-unsafe and does the bucketing after value collection.
+    Over time, I think that there would be potential to merge the two Histogram classes.
+*   `CounterFamily`, `GaugeFamily`, `HistogramFamily`: An interface for obtaining `Counter`/`Gauge`/`Histogram` instances that share a common purpose and only differ by a set of labels.
```

Comment:
The purpose of the "families" is not completely clear to me from this sentence. Looking at the [concrete example](https://github.com/stevewolter/cartographer/blob/6985ed9c68020986247afa211780cca39baaa3b7/cartographer/mapping_3d/pose_graph/constraint_builder.cc#L46) in your prototype they are used as a sort of collection for the three constraint builder histograms. Why should they only differ by the labels, and not e.g. in the boundaries?

Reply:
Yes, I also find it difficult to put the distinction into words. May I just throw the long description at you and you point out what you find important from a cartographer dev's view, and if it is clearer?

A family is essentially an aggregation domain. All metrics in the same family are "about the same thing" and can be aggregated safely. For example, a family might be "RPC  responses", with individual labels calling out RPC methods and RPC result classes. It's totally reasonable to just sum over all the labels in the family to get the total number of RPCs. For metrics in different families, this is generally not true - it's apples and oranges. To enable the relationship, the family shares the same type and the same metric name.

Histograms are the most complex case here, because of the question of boundaries. Histograms are an afterthought in varz (Prometheus' ancestor), and you still notice. Technically, you could set the boundaries differently for each individual histogram, but then you lose the easy aggregation property :-(.

Reply:
Ahh thanks for the long description, I think I get the point now. To put it in simple words how I see it: a family is an aggregate of metrics that "makes sense", i.e. only contains instances that can be compared / summed / etc because they share the same (e.g. mathematical) properties.

However, when I look at your example code: isn't putting the `rotational_score` and the `low_resolution_score` into the same family a violation of this pattern? Although I'm not so familiar with the 3D constraint builder, they sound like different metrics. At least *inside* the family, we would be dealing with apples and oranges.

Reply:
Boundaries need to match so they can be aggregated.
As an example, fixed-size bucket counts of infra and inter constraint scores can be aggregated to constraint scores (if the bucket ranges match).